File: java/src/priorityQueue/DisJointSetTreeHeuristics.java
package priorityQueue;

public class DisJointSetTreeHeuristics {
    private int[] parent = new int[6];

    private int[] rank = new int[6];

    public String toString() {
        StringBuffer sb = new StringBuffer();
        for(int i = 0 ; i < this.parent.length ; i++) {
            sb.append(this.parent[i]);
            sb.append("|");
            sb.append(this.rank[i]);
            sb.append(",");
        }

        return sb.toString();
    }

    public void makeSet(int i) {
        if (i < 0 || i >= this.parent.length)
            throw new IndexOutOfBoundsException();

        this.parent[i] = i;
    }

    /* path comparision heuristic */
    public int find(int i) {
        if (i < 0 || i >= this.parent.length)
            throw new IndexOutOfBoundsException();

        if (i != this.parent[i])
            this.parent[i] = this.find(this.parent[i]);

        return this.parent[i];
    }

    /* union by rank heuristic */
    public void union(int i, int j) {
        int i_id = this.find(i);
        int j_id = this.find(j);

        if (this.rank[i_id] > this.rank[j_id]) {
            this.parent[j_id] = i_id;
        } else {
            this.parent[i_id] = j_id;
            if (this.rank[i_id] == this.rank[j_id])
                this.rank[j_id] += 1;
        }
    }
}

File: python/priorityQueue/disJointSetTreeHeuristics.py
class DisJointSetTreeHeuristics:
    def __init__(self):
        self.parent = [None] * 6
        self.rank = [None] * 6

    def toString(self):
        return ','.join([str(i) for i in self.parent]) + '|' + ','.join([str(i) for i in self.rank])

    def makeSet(self, i):
        if (i < 0 or i >= len(self.parent)):
            raise IndexError
        self.parent[i] = i
        self.rank[i] = 0

    def find(self, i):
        if (i < 0 or i >= len(self.parent)):
            raise IndexError
        if (i != self.parent[i]):
            self.parent[i] = self.find(self.parent[i])

        return self.parent[i]

    def union(self, i, j):
        i_id = self.find(i)
        j_id = self.find(j)

        if (self.rank[i_id] > self.rank[j_id]):
            self.parent[j_id] = i_id
        else:
            self.parent[i_id] = j_id
            if (self.rank[i_id] == self.rank[j_id]):
                self.rank[j_id] += 1

